src/server/streamableHttp.ts

```diff
@@ -0,0 +1,368 @@
+        ('method' in msg && !('id' in msg)) || ('result' in msg || 'error' in msg));
+
+      if (hasOnlyNotificationsOrResponses) {
+        // if it only contains notifications or responses, return 202
+        res.writeHead(202).end();
+
+        // handle each message
+        for (const message of messages) {
+          this.onmessage?.(message);
+        }
+      } else if (hasRequests) {
+        const headers: Record<string, string> = {
+          ...this._customHeaders,
+          "Content-Type": "text/event-stream",
+          "Cache-Control": "no-cache",
+          Connection: "keep-alive",
+        };
+
+        // For initialization requests, always include the session ID if we have one
+        // even if we're in stateless mode
+        if (this._sessionId !== undefined) {
+          headers["mcp-session-id"] = this._sessionId;
+        }
+
+        res.writeHead(200, headers);
+
+        // Store the response for this request to send messages back through this connection
+        // We need to track by request ID to maintain the connection
+        for (const message of messages) {
+          if ('method' in message && 'id' in message) {
+            this._sseResponseMapping.set(message.id, res);
+          }
+        }
+
+        // handle each message
+        for (const message of messages) {
+          this.onmessage?.(message);
+        }
+        // The server SHOULD NOT close the SSE stream before sending all JSON-RPC responses
+        // This will be handled by the send() method when responses are ready
+      }
+    } catch (error) {
+      // return JSON-RPC formatted error
+      res.writeHead(400).end(JSON.stringify({
+        jsonrpc: "2.0",
+        error: {
+          code: -32700,
+          message: "Parse error",
+          data: String(error)
+        },
+        id: null
+      }));
+      this.onerror?.(error as Error);
+    }
+  }
+
+  /**
+   * Handles DELETE requests to terminate sessions
+   */
+  private async handleDeleteRequest(req: IncomingMessage, res: ServerResponse): Promise<void> {
+    if (!this.validateSession(req, res)) {
+      return;
+    }
+    await this.close();
+    res.writeHead(200).end();
+  }
+
+  /**
+   * Validates session ID for non-initialization requests when session management is enabled
+   * Returns true if the session is valid, false otherwise
+   */
+  private validateSession(req: IncomingMessage, res: ServerResponse): boolean {
+    const sessionId = req.headers["mcp-session-id"];
+
+    if (!sessionId) {
+      // Non-initialization requests without a session ID should return 400 Bad Request
+      res.writeHead(400, this._customHeaders).end(JSON.stringify({
+        jsonrpc: "2.0",
+        error: {
+          code: -32000,
+          message: "Bad Request: Mcp-Session-Id header is required"
+        },
+        id: null
+      }));
+      return false;
+    } else if ((Array.isArray(sessionId) ? sessionId[0] : sessionId) !== this._sessionId) {
```

Comment:
I would just assert that this header value isn't an array (i.e., throw an exception if it is).